contribute/design/1.x/upgrade/volume-pools-upgrade.md

```diff
@@ -0,0 +1,265 @@
+the shell scripts to perform the upgrade tasks. 
+
+The upgrade scripts will be packaged into upgrade container, with an 
+entrypoint script that will invoke the right scripts depending on 
+the object being upgraded. 
+
+A custom resource called `UpgradeTask` will be defined with the 
+details of the object being upgraded, and will in trun be updated
+with the status of the upgrade by the upgrade-container scripts.  
+
+OpenEBS team will release an openebs-upgrade-tools container 
+that contains all the upgrade scripts supported by openebs. 
+
+The workflow followed by administrator is as follows:
+- Define the Custom Resource and a Service account
+- Create a UpgradeTask CR for the object being upgraded 
+- Create a Kubernetes Job with upgrade-container and passing the 
+  UpgradeTask details. 
+- Administrator can query the UpgradeTask to check the status 
+  and result of the upgrade.
+
+### Design Choices/Decisions
+
+This section captures some of the alternative design considered
+and the reasoning behind selecting a certain approach. 
+
+- A generic UpgradeTask CR vs task specific ones like JivaUpgradeTask,
+  CStorVolumeUpgradeTask, etc. Having specific tasks has the following
+  advantages:
+  * Each tasks can have its own spec. the fields may vary depending
+    on the resource being upgraded.
+  * Writing specific operators for each upgrade that operates on a
+    given type of task will make the upgrades more modular. 
+  
+  However, this also means that every time a new resource type is
+  added, another CR needs to be introduced, managed and learned by 
+  the user. This may still be ok. But a similar pattern where 
+  different specs are required is already addressed by the PVC. To 
+  keep the resources management at a minimum, the PVC type, sub-resource
+  spec pattern will be used to specifiy different resources under
+  a generic UpgradeTask CR. 
+
+  Note that, selecting the Generic Task - doesn't preclude from 
+  writing specific upgrade task operators and upgrades task CRs as long
+  as the management of these specific upgrade tasks are completely 
+  management by the operators and user need not interact with them. One
+  possible future implementation could be: The upgrade-operator can 
+  look at the upgrade task (consider it as a claim) and create a 
+  specific upgrade task and bind it. Then the specific upgrade 
+  operator will operate on the their own resources.  
+
+
+### High Level Design
+
+#### UpgradeTask CR Example
+
+Here is an example of UpgradeTask for upgrading Jiva Volume. The 
+conditions are added by the Upgrade Job
+
+```
+apiVersion: openebs.io/v1alpha1
+kind: UpgradeTask
+metadata:
+  name: upgrade-jiva-pv-001
+  namespace: openebs
+spec:
+  fromVersion: 0.9.0
+  toVersion: 1.0.0
+  jivaVolume:
+    pvName: pvc-3d290e5f-7ada-11e9-b8a5-54e1ad4a9dd4
+status:
+  phase: #INIT, STARTED and ERROR can be the supported phases
+  # conditions represent current reconciliation
+  # activities
+  conditions:
```

Comment:
Will the conditions be list containing all the stages the pvc went through or only a single condition which will change  its value as upgrade proceeds ?

Reply:
This will be list of conditions - which are now renamed to `upgradeDetailedStatuses`.